hr_timesheet_sheet/models/hr_timesheet_sheet.py

```diff
@@ -414,7 +416,7 @@ def action_timesheet_refuse(self):
     def button_add_line(self):
         for rec in self:
             if rec.state == 'draft':
-                rec.add_line()
+                rec.add_line(rec.add_line_project_id, rec.add_line_task_id)
```

Comment:
Do we really need to pass these as params? What's be benefit of that?

Reply:
Optional inheritance. Just in case you want to make another module that depends of this one, and wants other functionalities, then it's better to have options like this, to pass other projects or tasks.

Reply:
It seems to me that w/o parameters it's much easier to extend just add_line, otherwise I'd have to override both `button_add_line` and `add_line` if I'd like to add extra `add_line_smth`

Reply:
I differ with you. I think with parameters it enables more possible customization. If not, `_prepare_empty_analytic_line` method is restricted by what happens in `button_add_line` method, but  `add_line` method cannot influence.

Reply:
Got it, makes sense, can we do something like _preapare_new_line_params?

Reply:
I don't understand. What method are you mentioning?

Reply:
```

    @api.multi
    def button_add_line(self):
        for rec in self:
            if rec.state == 'draft':
                rec.add_line(rec.project_id)
                rec.reset_add_line()
        return True

def add_line(self, project_id, ):
        if self.add_line_project_id:
            values = self._prepare_empty_analytic_line(project_id)
            name_line = self._get_new_line_name(values)
    
```

Something like this

Reply:
Why: https://github.com/OCA/hr-timesheet/pull/154/files#diff-c470854eb5351f22cc8515bd7cfe525eR67
https://github.com/OCA/hr-timesheet/pull/154/files#diff-c470854eb5351f22cc8515bd7cfe525eR80

Reply:
Don't understand. If in `add_line` method I do `name_line = self._get_new_line_name()` instead of `name_line = self._get_line_name(project, task)`, that would go against the intention of giving more possible customization.

Reply:
`self._get_new_line_name(values)`?

Reply:
Anyhow, it's not related to this improvement and should be discussed separately, I guess.

Reply:
Let me try to re-express my vision: I'm trying to limit number of places that pass project + task to functions, in order to be able to extend project + task + something w/o need to inject into multiple places.

Reply:
`values` is a dictionary. I know what you are trying, but your experiments until now are not good. So, I think the best way to proceed is to try to merge this PR as it is now and later if you want you can make another with your thing, because if not, it would seem that this thing you are trying is really blocking this PR.

Reply:
Not blocking at all, I'll just review the commits to be portable to v12, and approve :) Totally agree on route